From: Christoph M. Becker <cmbecker69@gmx.de>
Date: Wed Jan 9 14:26:18 2019 +0100
Subject: [PATCH] Fix #77431 SplFileInfo::__construct() accepts NUL bytes

`SplFileInfo::__construct()` has to expect a path instead of a string,
analogous to `SplFileObject::__construct()`.

5.5 backport: https://github.com/shivammathur/php-src-backports/commit/672ba7aa6155fe5d423fd67da77ccb1bd0a239bd
---
diff --git a/ext/spl/spl_directory.c b/ext/spl/spl_directory.c
--- a/ext/spl/spl_directory.c
+++ b/ext/spl/spl_directory.c
@@ -1134,7 +1134,7 @@ SPL_METHOD(SplFileInfo, __construct)
 
 	zend_replace_error_handling(EH_THROW, spl_ce_RuntimeException, &error_handling TSRMLS_CC);
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &path, &len) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "p", &path, &len) == FAILURE) {
 		zend_restore_error_handling(&error_handling TSRMLS_CC);
 		return;
 	}
diff --git a/ext/spl/tests/bug77431.phpt b/ext/spl/tests/bug77431.phpt
new file mode 100644
--- /dev/null
+++ b/ext/spl/tests/bug77431.phpt
@@ -0,0 +1,9 @@
+--TEST--
+Bug #77431 (SplFileInfo::__construct() accepts NUL bytes)
+--FILE--
+<?php
+new SplFileInfo("bad\0good");
+?>
+--EXPECTF--
+Fatal error: Uncaught exception 'RuntimeException' with message 'SplFileInfo::__construct() expects parameter 1 to be a valid path, string given' in %s:%d
+Stack trace:%A
\ No newline at end of file
